api/v1_projects.inc

```diff
@@ -656,3 +657,47 @@ function _get_enabled_filter($query_params)
         return $bool;
     }
 }
+
+//---------------------------------------------------------------------------
+// Proofreading functions
+
+// checkout a page
+function api_v1_project_checkout($method, $data, array $query_params)
+{
+    $proof_project = new ProofProject($data[":projectid"], $query_params);
+    return $proof_project->checkout();
+}
+
+function api_v1_project_validatetext($method, $data, array $query_params)
+{
+    $text_checker = new TextChecker($data[":projectid"]);
+    return $text_checker->analyse();
+}
+
+function api_v1_project_page($method, $data, $query_params)
+{
+    $proof_project_page = new ProofProjectPage($data[":projectid"], $data[":pagename"], $query_params);
```

Comment:
We should not pass opaque arrays down into other functions, particularly class constructors. It's the API's responsibility as the front-line to the user to validate API input are valid values -- they could be the _wrong_ values (it's a valid state but the wrong state) and the underlying code could decide it's wrong and raise something, but the API should sanitize user data.

Reply:
I am inclined to move the project state and page state into the URI. I'll do this in a separate commit.

Reply:
I think having them in the request body is the right place. Having a resource locator that looks like `v1/projects/:projectid/pages/:pagename/state/:state/` where there is only one legal value for state doesn't really make any sense -- ditto project state. We should just pull them out in the api code and not here.

Reply:
I see what you mean. I am quite happy to make it work however you like.
One could make sense of this by saying that the project can be in different states at different times and making a request for a resource which is a project in a given state will either succeed or fail in the sense that the resource does not exist.

Reply:
I think your URL schema for this is good -- I would just pull all of the ProofProjectPage constructor's parameters out of $query_params here in the API rather than in the class itself.

Reply:
I'll do that.

Reply:
Done this now.